src/sentry/ingest/transaction_clusterer/tree.py

```diff
@@ -0,0 +1,136 @@
+
+For example, a blog project might contain transaction names along the lines of:
+
+/users/my-user-name/posts/2022-01-01-that-one-time-i-did-something
+
+We build a tree that looks like this:
+
+/users
+  /my-user-name
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+  /my-user-name2
+    /posts
+      /2022-12-31-i-did-something-too
+  /my-user-name3
+    /posts
+      /2022-07-15-but-what-about-me
+    /settings
+
+As soon as the node /users has more than X children, we merge them:
+
+/users
+  /*
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+      /2022-12-31-i-did-something-too
+      /2022-07-15-but-what-about-me
+    /settings
+
+If /posts now also has more than X children, those are merged as well:
+
+/users
+  /*
+    /posts
+      *
+    /settings
+
+Resulting in the following replacement rules:
+
+/users/*/posts/*  # To replace both identifiers
+/users/*          # For URLs with only one identifier, e.g. /users/*/settings
+
+"""
+
+from collections import UserDict, defaultdict
+from typing import Iterable, List, Optional, Union
+
+from typing_extensions import TypeAlias
+
+from .base import Clusterer, ReplacementRule
+
+__all__ = ["TreeClusterer"]
+
+
+class Merged:
+    pass
+
+
+#: Symbol representing a merged node
+MERGED = Merged()
+
+#: Separator by which we build the tree
+SEP = "/"
+
+
+# TODO: Metrics for tree size, spans for merge, etc.
+
+
+class TreeClusterer(Clusterer):
+    def __init__(self, *, merge_threshold: int) -> None:
+        self._merge_threshold = merge_threshold
+        self._graph = Node()
+
+    def add_input(self, transaction_names: Iterable[str]) -> None:
+        for tx_name in transaction_names:
+            parts = tx_name.split(SEP)
+            node = self._graph
+            for part in parts:
+                node = node.setdefault(part, Node())
+
+    def get_rules(self) -> List[ReplacementRule]:
+        """Merge high-cardinality nodes in the graph and extract rules"""
+        self._graph.merge(self._merge_threshold)
+        # Generate exactly 1 rule for every merge
+        rule_paths = [path for path in self._graph.paths() if path[-1] is MERGED]
+
+        # TODO: Warn if top level collapse
+
+        # Sort by path length, descending (most specific rule first)
+        rule_paths.sort(key=len, reverse=True)
+
+        return [self._build_rule(path) for path in rule_paths]
+
+    @staticmethod
+    def _build_rule(path: List["Edge"]) -> ReplacementRule:
+        return ReplacementRule(SEP.join(["*" if isinstance(key, Merged) else key for key in path]))
+
+
+#: Represents the edges between graph nodes. These edges serve as keys in the
+#: node dictionary.
+Edge: TypeAlias = Union[str, Merged]
+
+
+class Node(UserDict):  # type: ignore
+    """Keys in this dict are names of the children"""
+
+    def paths(self, ancestors: Optional[List[Edge]] = None) -> Iterable[List[Edge]]:
+        """Collect all paths and subpaths through the graph"""
```

Comment:
IIUC, the subpaths are to generate the `*`s above. I think we can simplify this to something like the following:

```python
def __init__(...):
    self.path: str = 'whatever'
    self.children: dict[str, Node] = {}  # mapping from token or path to another node
    self.terminal: bool = false  # terminal nodes are where there's an existing path ending here
    self.merged: bool = false  # if this node should be merged
```
This code could also look better with the `merge` logic properly done.

Then the logic to get the paths could look something like this:

```python
if self.terminal and self.merged:
    return '**'
if self.merged:
    return '*'
return self.path
```

Reply:
Not sure if I understand the `**` vs `*` reasoning, but that might be because we have not specified how Relay should interpret these rules.

My interpretation so far is that Relay should prefix-match `/a/b/*`, so it would match both `/a/b/c` and `/a/b/c/d`, but in both cases, it would replace only `c`.

Reply:
These are globs in relay. This means `/a/b/*` matches `/a/b/c` but not `/a/b/c/d`, see [online](https://www.glob101.com/view/yo4N0DpwFpFjvV5Zh3Xs), and [struct](https://github.com/getsentry/relay/blob/ab46f30a4fcf3682f60cdc6f57a1d16f21c7f7cc/relay-common/src/utils.rs#L14) in relay with [sample test](https://github.com/getsentry/relay/blob/ab46f30a4fcf3682f60cdc6f57a1d16f21c7f7cc/relay-common/src/utils.rs#L163-L165).

If we want a rule to match both, that'd be `/a/b/*/**` after adding a `/` to every path: `/a/b/c/` and `/a/b/c/d/` ([see](https://www.glob101.com/view/BBYATGPlgvuimtEJV04R)). Another alternative is to generate two rules: `/a/b/*` and `/a/b/*/**`; this option doesn't require any modifications to the path, but it's 2x rules.

I've assumed that we're only replacing `*` and not `**`.

In the example in the docs at the top of this file, the resulting rules would be the following:
```
/users/*/posts/*/**  # replaces two identifiers
/users/*/**  # replaces one identifier
```

Reply:
Assuming that we're going for the `/*/**` solution, I changed the rule generation accordingly.